### PR TITLE
upgrade Eigen to 3.3.7, fix a few warnings

### PR DESCRIPTION
**Pull request by redpony**
Recent versions of Eigen have much better support for modern CPU's SIMD operations. This patches AD3 to use the most recent version (tested on a few problems).

### CODE REVIEW

File: Eigen/Eigen [hunk 1]
```diff
@@ -1,2 +1,2 @@
 #include "Dense"
-//#include "Sparse"
+#include "Sparse"
```
Comment (vene) [inline Eigen/Eigen]: @andre-martins was the Sparse module intentionally commented out here in your original code to reduce build time or something similar? Or is this just the way the Eigen release looked?

**Comment by redpony** [comment]
This is just the way the Eigen release looked. I didn't do anything to it.

File: Eigen/src/Core/util/Macros.h [hunk 2]
```diff
@@ -12,94 +12,461 @@
 #define EIGEN_MACROS_H
 
 #define EIGEN_WORLD_VERSION 3
-#define EIGEN_MAJOR_VERSION 1
-#define EIGEN_MINOR_VERSION 1
+#define EIGEN_MAJOR_VERSION 3
+#define EIGEN_MINOR_VERSION 7
```
Comment (vene) [inline Eigen/src/Core/util/Macros.h]: for reference for the changelog, the old version was 1.1

**Comment by vene** [comment]
@redpony the question was for @andre-martins :stuck_out_tongue: the way it looks in this PR looks good to me, the way it was before struck me as weird.

I triggered travis tests on my branch and all looks good: https://travis-ci.org/vene/AD3/builds/477943758

I'll merge after @andre-martins confirms whether commenting out `Eigen/Sparse` was intentional.

File: Eigen/Eigen (hunk 1 above)
Comment (andre-martins) [reply Eigen/Eigen]: Thanks @redpony ! @vene I don't think I edited anything in the Eigen code, this was probably commented in the older Eigen release.

**Comment by vene** [comment]
Great! Thanks a lot @redpony! Merging.

Now, the warnings from ad3 itself stand out. I should try to find some time to fix them..

**Pull request merged**